Introduce memory management on cluster link buffers

ny0312 opened the pull request:
This PR introduced memory management on cluster bus link send buffers, to address issues raised in https://github.com/redis/redis/issues/9688.

The interface changes are (need core-team consensus):
 * Introduced a new `cluster-link-sendbuf-limit` config that caps memory usage of cluster bus link send buffers. 
   * Its default is chosen to be 0 (aka infinite).
 * Introduced a new `CLUSTER LINKS` command that displays current TCP links to/from peers as a RESP3 map.
 * Added `mem_cluster_links` field under the `Memory` section of `INFO` command output, which displays the overall memory usage by all current cluster links. 
   * Value of `mem_cluster_links` is also deducted from `used_memory_dataset` to better isolate the memory usage of real key values.
   * Memory usage of cluster links are still counted toward the `maxmemory` limit, so over consumption by cluster links can still cause evictions or `OOM` errors (no behavior change).
 * Added `total_cluster_links_buffer_limit_exceeded` field under the `CLUSTER INFO` command output, which displays the accumulated count of cluster links freed due to `cluster-link-sendbuf-limit`.


On an implementation level:
 * In `clusterNode` struct, added an `inbound_link` field pointing to the inbound link accepted from this `node`. Before this change, inbound links are not associated with the corresponding `clusterNode` and are loosely managed.

ny0312 commented:
This is a new PR that replaces old PR: https://github.com/ny0312/redis/pull/1

There were some comments left on the old PR that are now addressed in this new PR.

madolson commented on tests/cluster/tests/24-links.tcl:
```diff
@@ -0,0 +1,99 @@
+    # Revive primary2
+    exec kill -SIGCONT $primary2_pid
+
+    # Reset configs on primary1 so config changes don't leak out to other tests
+    catch {$primary1 CONFIG set cluster-node-timeout $oldtimeout}
+    catch {$primary1 CONFIG set cluster-link-sendbuf-limit $oldlimit}
```
Why are we catching these exceptions? Also if the test fails midway through these changes will get lost.

ny0312 replied:
Good question. I was copying what existing tests did. Looking at it now, it makes more sense to let the exceptions be raised and fail the test, instead of catching and dropping them.

Re: tests failing midway through. If one test case fail, does the test framework continue to run the remaining tests? If yes, this sounds like the generic issue, i.e. the lack of isolation between cluster tests. How do existing tests handle prevent "leaks"?

madolson replied:
Yeah, the testing framework is a little bit silly in that it doesn't do a full reset of the cluster nodes.

madolson reviewed (commented):
Still some thoughts on the other PR, but this one seems OK to me. We also need a change in redis.conf to describe the parameter.

madolson commented on src/cluster.c:
```diff
@@ -4599,6 +4738,9 @@ void clusterCommand(client *c) {
 "    start, end, master and replicas IP addresses, ports and ids",
+"LINKS",
+"    Return information about all network links between this node and its peers. Output format:",
+"    <fd > <to|from> node=<peer node id> ctime=<creation time in milliseconds> events=<r><w> sbufalloc=<send buffer allocated size in bytes> sbufused=<send buffer used size in bytes>",
```
This isn't accurate anymore.

ny0312 replied:
Fixed.

madolson reviewed (commented):
LGTM otherwise.

madolson commented:
@redis/core-team Please review, there are two core team decisions. The first is the config to limit maximum amount of outbound data, the second is the command to describe the information for the cluster links.

oranagra commented on redis.conf:
```diff
@@ -1574,6 +1574,13 @@ lua-time-limit 5000
+# this limit. This is to primarily prevent send buffers from growing unbounded on links
+# toward slow peers (E.g. PubSub messages being piled up).
+#
+# cluster-link-sendbuf-limit 1gb
```
so this is similar to the client output buffer limit, right?
dropping connections when they consume too much memory.
which is suppose to be the case only when a cluster node hangs (or on a burst of local PUBLISH calls)?.

I see that there too the limit is enabled by default for some types.
specifically, the replica client are limited to 256mb, and the pubsub clients to 32mb.
i assume that considering that there are usually not a lot of replicas, this makes the cluster nodes more similar to pubsub clients, so maybe we should aim much lower with the default limit?

for clients, we now also have a limit for the sum memory of all of them (`maxmemory-clients`) but i suppose that for cluster nodes we don't need that (unlike clients, their number is bound and controlled by an admin)

madolson replied:
I forgot where we came up with 1GB, but lowering seems reasonable. I think the limit is somewhere in between the two. We expect the other side to be consuming the data as fast as possible, like a replica, but it also has the high write amplification like pubsub. Since each individual link is small, I agree that 32mb sounds reasonable.

ny0312 replied:
This feature aims to protect against any case where a peer cannot drain cluster link send buffers as fast as Redis can fill them, e.g. hang peer, slow peer, etc.

I chose 1GB as the default limit based on the max possible size of a single PubSub command, which is 1GB by default (`client-query-buffer-limit`'s default value). I wanted cluster link buffer to be able to fit at least a single PubSub command by default.

Let me know what you think.

oranagra replied:
i suppose that effectively this means for many deployments as if the default is practically unlimited (will still cause mass eviction / OOM).
maybe that's a good thing, to consider that as if it's not a "breaking change", i.e. the new default has the same behavior as the old versions, but if we look at it like that, maybe just set it to 0 by default?

if we have any specific reasoning like the one you just mentioned, maybe we should document it in the redis.conf and also in the top comment.

ny0312 replied:
I dont know if I agree that the default of 1GB is practically unlimited. 1GB is indeed a generous limit but it still offers some protection which is better than none, because slow peers due to network issues are real issues.

I believe the protection offered by this new limit should be turned on by default. I would even prefer lowering it to 32MB than `0`, but IMO 1GB makes sense because my reasoning in the previous post.

Added this reasoning to `redis.conf` and top comment.

madolson replied:
I suppose one problem with 32MB is that pubsub messages might get constantly dropped and it would be very hard to notice. Unlike a regular pubsub client, which will get repeatedly disconnected if large messages are put in it, the cluster link will just get dropped and re-established, but the "connected" pubsub client won't notice anything. So I guess I'm back in the 1gb camp, and we'll use it more as a mechanism to make sure it doesn't grow unbounded.

oranagra replied:
I was arguing that for **some** deployments it's practically unlimited. I.e. Ones with not a lot of memory (either physical, or just the maxmemory config), or ones with a lot of cluster nodes. 

So considering that, i was arguing that maybe the default should be disabled, and thus not changing default behaviour on upgrades. 

P.s. Note that nearly all the memory limit setting in Redis are off by default (except for client buffers). This is because it's impossible to come up with a number that will be a good default for everyone. 

Anyway, considering what @madolson wrote, about getting disconnections and reconnection, resulting the silent loss of messages with no clear issue indicated anywhere, I'm not sure i like that feature at all..

ny0312 replied:
I hear you about it's impossible to pick a default to make everyone happy.

I also agree that **silently** dropping cluster links and PubSub broadcast messages is bad.

Please allow me to re-visit why I wrote this PR in the first place. Before this PR, cluster link buffers can grow unbounded and:
1. It's impossible to observe if it is happening
2. It's impossible to isolate which link is using memory
3. It's impossible to mitigate it with minimal impact on the server (there is no easy way to free a specific cluster link that is over-using memory).

In this PR:
1. `mem_cluster_links` INFO field is introduced to address (1)
2. `CLUSTER LINKS` command is introduced to address (2)
3. `cluster-link-sendbuf-limit` config is introduced to address (3).

In this context, re: `enabling cluster-link-sendbuf-limit by default or not`. I was leaning toward enabling by default. Because I trust the default of 1GB will not produce "false alarms". That is, we never expect a cluster link to use >1GB by default. When a cluster link does, it means Redis is operating in an abnormal and degraded mode, which we want to get out of. Enabling by default just means Redis would auto-mitigate such an abnormal situation, which is a net positive. On smaller deployments where there is not a lot of memory, this auto-mitigation will never take effect unfortunately, so Redis regresses to its current behavior, but at least it doesn't make things worse.

Re: `Using 1GB as default`. I guess my reasonings can be summarized as:
1. In conjunction with `client-query-buffer-limit'`s default value of 1GB, a default 1GB of `cluster-link-sendbuf-limit` should/would not produce false alarms, which is important.
2. While it's impossible to pick a default everyone likes, sometimes we still have to pick a sensible number to give us some bottom line protection. E.g. `client-query-buffer-limit` default to 1GB is a good example. I tried to follow that.

Re: `silently dropping cluster links and PubSub messages`. I agree this is bad. This is a separate issue. It seems like whenever `cluster-link-sendbuf-limit` takes effect and frees a link, we also want to print a log and/or increment a metric count (e.g. a `cluster_links_freed_oom` INFO field, etc). So it's not "silent". I can add that in the next revision if we all agree to if we want a log, a metric, or both.

Thank you for your feedback. Please let me know your thoughts.

madolson replied:
I'm still in favor of setting 1gb, but only making this available in Redis 7. In some sense it is a backward breaking change, but it is bounding the problem moving forward.

madolson replied:
Decision by core team was to set the default to 0.

ny0312 replied:
Done.

oranagra commented on src/cluster.c:
```diff
@@ -4390,6 +4476,80 @@ sds clusterGenNodesDescription(int filter, int use_pport) {
+        }
+    }
+    dictReleaseIterator(di);
+    addReplyArrayLen(c, num_links);
```
i suppose we better use differed reply, rather than run the loop twice.

ny0312 replied:
Changed. Didn't know Deferred Reply is a thing.

oranagra commented on src/cluster.c:
```diff
@@ -4390,6 +4476,80 @@ sds clusterGenNodesDescription(int filter, int use_pport) {
+void addReplyClusterLinkDescription(client *c, clusterLink *link) {
+    addReplyMapLen(c, 7);
+
+    addReplyBulkCString(c, "fd");
+    addReplyLongLong(c, link->conn->fd);
```
is it really needed to expose the fd to users?

madolson replied:
For some reason I thought we also exposed the FD in client list, but we don't. I would agree that I don't think we should emit it.

ny0312 replied:
Removed `fd` from the `cluster link` output.

oranagra commented on src/server.c:
```diff
@@ -6343,6 +6347,7 @@ sds genRedisInfoString(const char *section) {
             "mem_total_replication_buffers:%zu\r\n"
             "mem_clients_slaves:%zu\r\n"
             "mem_clients_normal:%zu\r\n"
+            "mem_cluster_links:%zu\r\n"
```
this is another interface change

ny0312 replied:
Yes. In general, IMO memory consumption on the cluster bus has been a blind spot. 

For example, before we had [this commit](https://github.com/redis/redis/commit/8647addf934c453e933bb5f26ab4950fd5924a04), cluster link buffers never shrunk and could grow unbounded, but we have no way to observe/monitor it while it was happening.

oranagra replied:
by "interface change" i meant it should be listed in the top comment, for the purpose of release notes and approval of reviewers that don't have time to read the code.

i.e. all interface changes are "major decisions" that must have lazy consensus by the core-team.

ny0312 replied:
I see. Good to know. Thanks for the info.

Added to the top comment.

oranagra reviewed (commented):
i didn't review the majority of the code. just the parts related to the new interfaces.
i suppose the feature is valid, but i'd like to argue about the default value for the config.

madolson commented on src/cluster.c:
```diff
@@ -3581,20 +3627,49 @@ int clusterNodeCronHandleReconnect(clusterNode *node, mstime_t handshake_timeout
+
+static void freeClusterLinkOnBufferLimitReached(clusterLink *link) {
+    if (link != NULL && sdsalloc(link->sndbuf) > server.cluster_link_sendbuf_limit_bytes) {
+        freeClusterLink(link);
```
I suppose we should add a log message here to explain what happened?

ny0312 replied:
Added a WARNING log message, considering this is not expected in normal operation.

madolson commented:
@ny0312 The decision was that we will set the default to 0, implying infinite today, and we'll evaluate how it works. We'll document that you can do something about it.

lkerzanml commented:
谢谢

madolson commented on src/server.c:
```diff
@@ -6845,6 +6851,12 @@ sds genRedisInfoString(const char *section) {
         "# Cluster\r\n"
         "cluster_enabled:%d\r\n",
         server.cluster_enabled);
+
+        if (server.cluster_enabled) {
+            info = sdscatprintf(info,
+                    "total_cluster_links_buffer_limit_exceeded:%llu\r\n",
+                    server.cluster->stat_cluster_links_buffer_limit_exceeded);
+        }
```
Let's move this into the `cluster info` command.

ny0312 replied:
Done

ny0312 commented:
Accompanying doc PR: https://github.com/redis/redis-doc/pull/1710

Thanks.

madolson reviewed (approved):
LGTM, @redis/core-team back to ya'll for approval.

oranagra commented on redis.conf:
```diff
@@ -1576,6 +1576,17 @@ lua-time-limit 5000
+# This limit is disabled by default. Enable this limit when 'mem_cluster_links' INFO field
+# and/or 'sendbufalloc' entries in the 'CLUSTER LINKS` command output continuously increase.
+# Minimum limit of 1gb is recommended so that cluster link buffer can fit in at least a single
+# PubSub message by default. (client-query-buffer-limit default value is 1gb)
```
the documented default, and the default value in the conf file are outdated (the one in config.c is ok)

madolson replied:
While I'm here I guess I'll fix this.

oranagra reviewed (commented):
one comment. otherwise LTTM.

oranagra commented:
The top comment of the PR is outdated too.

oranagra reviewed (approved):
LGTM

madolson commented:
@ny0312 Sorry to ask more from you, but we moved to a new command definition system:
```
If you had any PR that modifies the command table, you'll need to rebase, modify the json files, and run the python script that generates commands.c (and commit that too).
Same thing about modifying some things about the command (like argument, response, etc. they're all documented there now).

If you have any redis-doc PR that changes commands.json, make sure that the change is reflected in the json files in the redis repo, the command.json in redis-doc is now in "code" freeze (for now you can keep the change in both repos, but make sure they're "Identical"), we'll soon delete the json file from redis-doc and take the data from redis in some way.
```

You can transfer the contents of the doc PR over to this CR now. 

@yossigo Can you approve the major changes if you have time, so we can get this merged?

oranagra commented:
@ny0312 FYI, i'd prefer to avoid rebase, amend, and force-push, it's harder to review what changed.
since we're merging most PRs with a squash-merge, it doesn't matter how many incremental commits they contain.
in some cases we have PRs that we plan to merge without squash (if they contain several topics, or some mass refactory in a separate commit), and in these cases, when we rebase, we must do a separate force-push (not mixed with any actual change), so we can let GH show us the diff that each force-push introduced.

anyway, long story short, does the last push contain any actual changes? or just a rebase?

yossigo reviewed (approved):
LGTM, no complete code review but I agree with the interface change and concept.

madolson commented:
@oranagra This is just a rebase to handle the merge conflicts because there is the new command definitions.

oranagra commented:
the new CLUSTER LINKS command will be implicitly supported in async-loading when this PR and #9878 are both individually applied, right? (no need to make any special changes in one for the benefit of the other)

p.s. don't forget to make a redis-doc PR.

ny0312 commented:
@oranagra My apologies. I'm clumsy with github PRs. I simply did not know that `we're merging most PRs with a squash-merge`. 

To answer your question, my last force-push contained:
1. rebase
2. Added a new `cluster-links.json` file under `src/commands/` folder.
3. Added a line to `commands.c` for `cluster links`, manually. Later I realize that `commands.c` is supposed to be auto-generated. So my next revision will remove my manual change to `commands.c` and replace it with an auto-generated change.

ny0312 commented:
@oranagra Yes, running `CLUSTER LINKS` commands during async loading should be fine. I gave it the same flags that `CLUSTER NODES` have.

The accompanying redis-doc PR is here: https://github.com/redis/redis-doc/pull/1710. I dont see a commit on the `redis-doc` repo that removes all existing json files. I dont want to start making `redis-doc/commands/*.json` and `redis/src/commands/*json` to diverge, so im stilll keeping the `cluster-links.json` file in my `redis-doc` PR.

oranagra commented:
For now we didn't yet get rid of commands.json in redis-doc, so we need to be very careful and make sure we do the same changes in both places, so we we eventually delete it, no data is lost

madolson commented:
@oranagra Can you verify that commands.c and the json file were implemented correctly? They look right to me but I haven't looked that much. Besides that I think this is ready to merge along with the doc PR.

oranagra reviewed (approved):
LGTM

sundb commented:
`Disconnect link when send buffer limit reached` test failed in my daily CI.
https://github.com/sundb/redis/runs/4560035116?check_suite_focus=true
https://github.com/sundb/redis/runs/4560034978?check_suite_focus=true

oranagra commented:
@ny0312 can you please look into these failures.
see also
https://github.com/redis/redis/runs/4654435937?check_suite_focus=true#step:10:1262
```
01:17:54> Each node has two links with each peer: FAILED: Expected 19*2 eq 37 (context: type eval line 11 cmd {assert {$num_peers*2 eq $num_links}} proc ::foreach_instance_id)
```

madolson commented:
@oranagra @ny0312 that issue specifically is caused because the cluster tests do hard resets in between tests, and the nodes may not have re-established all links, especially for slower tests. I was looking into the other failures a bit last night, and don't have any other suggested fixes at the moment.

oranagra commented:
some failure in the FreeBSD CI in `Disconnect link when send buffer limit reached`
https://github.com/redis/redis/runs/4681750680?check_suite_focus=true#step:4:7614
This may be an issue with the CI itself (it't not stable), but also, maybe there's some timing issue in the test.

madolson commented:
I'll ping @ny0312 on our AWS slack, I'm not sure he's noticed this.

oranagra commented:
@madolson a few days ago he told me he's on vacation 8-)

madolson commented:
I probably should have known that :D.

I was able to slowly validate it was this commit that is causing both the crashes on sanitizer and the buffer disconnect issues. I have no idea why though, it might just be we're dumping a bunch of memory into the test which doesn't seem to be done elsewhere in the cluster test.

ny0312 commented:
I'm back. I will sync up with Maddy and investigate and fix what is going on. Sorry and thanks.

oranagra commented:
any news?
i must admit i didn't notice it fail lately, but got this one now:
https://github.com/redis/redis/runs/4733591841?check_suite_focus=true
```
00:47:22> Disconnect link when send buffer limit reached: error writing "sock802fbc590": broken pipe
```

it could be just a random failure of the slow freebsd CI (we see broken pipes in random places).
note that in the past we saw:
```
00:44:36> Disconnect link when send buffer limit reached: error writing "sock80264c2d0": connection reset by peer
```
as well as other types of errors, on **non**-freebsd CI.
* a hung triggering a SIGSEGV kill by the test framework on a sanitizer run.
* a the following failure on centos+tls:
```
01:17:54> Each node has two links with each peer: FAILED: Expected 19*2 eq 37 (context: type eval line 11 cmd {assert {$num_peers*2 eq $num_links}} proc ::foreach_instance_id)
```

ny0312 commented:
Hey sorry for the late response. But I have been trying to reproduce this test failure, but couldn't.

This is the command I run (I copied it from https://github.com/redis/redis/runs/4654435988?check_suite_focus=true):

```
gmake || exit 1 ; if echo "" | grep -vq redis ; then ./runtest --verbose --timeout 2400 --no-latency --dump-logs  || exit 1 ; fi ; if echo "" | grep -vq modules ; then MAKE=gmake ./runtest-moduleapi --verbose --timeout 2400 --no-latency --dump-logs  || exit 1 ; fi ; if echo "" | grep -vq sentinel ; then ./runtest-sentinel  || exit 1 ; fi ; if echo "" | grep -vq cluster ; then ./runtest-cluster  || exit 1 ; fi ;
```

I've run it numerous times locally. I've run it numerous times on a FreeBSD 64 bit EC2. All with the latest `unstable` branch. Never able to reproduce the failure.

From the error message, it looks like Redis crashed during my test. The crash must be flaky and non-deterministic.

Any ideas that could help reproduce it?

Another question - how do I preserve redis logs from these test runs?

sundb commented:
@ny0312 It's bound to happen every time on my local freebsd vm, I've only allocated 1 core and 256M RAM to it, and since I don't know about clusters, there's no way to solve this problem directly, maybe I can package the logs and send them to you.

sundb commented:
When memory is exhausted, this test is bound to fail, and here are some logs after enabling `loglevel debug`.

```
16316:M 14 Jan 2022 12:48:51.257 . I/O error reading from node link: connection closed
16316:M 14 Jan 2022 12:48:51.315 . I/O error reading from node link: connection closed
16316:M 14 Jan 2022 12:48:51.315 . I/O error reading from node link: connection closed
16316:M 14 Jan 2022 12:48:51.318 . Unable to connect to Cluster Node [127.0.0.1]:40000 -> accept: Resource temporarily unavailable
16316:M 14 Jan 2022 12:48:51.318 . Unable to connect to Cluster Node [127.0.0.1]:40002 -> accept: Resource temporarily unavailable
16316:M 14 Jan 2022 12:48:51.318 . Unable to connect to Cluster Node [127.0.0.1]:40001 -> accept: Resource temporarily unavailable
```

```
16314:M 14 Jan 2022 12:48:51.254 . I/O error reading from node link: Connection reset by peer
16314:M 14 Jan 2022 12:48:51.257 . I/O error reading from node link: connection closed
```

ny0312 commented:
Huh, looks like file descriptors are exhausted?

You mentioned you reproduced this on a FreeBSD vm. Could you please what exact is your setup? Which VirtualMachine?

In the meanwhile, I will try to simulate your setup to reproduce. I will provision a EC2 with FreeBSD, 1 core and 256MB memory.

Also, if you could share the full Redis logs for the failed test, that'd be great. (For example you could attach a file here, or use something like https://justpaste.it/)

Thanks a lot!

sundb commented:
@ny0312 justpaste.it can't upload files, so I created a repository to hold the logs(https://github.com/sundb/log).

* system: `Freebsd 12 64bit`
* vm: VMware

* ulimit
```
-t: cpu time (seconds)              unlimited
-f: file size (blocks)              unlimited
-d: data seg size (kbytes)          33554432
-s: stack size (kbytes)             524288
-c: core file size (blocks)         unlimited
-m: resident set size (kbytes)      unlimited
-l: locked-in-memory size (kbytes)  unlimited
-u: processes                       10240
-n: file descriptors                10240
-b: socket buffer size (bytes)      unlimited
-v: virtual memory size (kbytes)    unlimited
-p: pseudo-terminals                unlimited
-w: swap size (kbytes)              unlimited
-k: kqueues                         unlimited
-o: umtx shared locks               unlimited
```

* change loglevel
```diff
diff --git a/tests/cluster/run.tcl b/tests/cluster/run.tcl
index c81d8f39d..ddf9986aa 100644
--- a/tests/cluster/run.tcl
+++ b/tests/cluster/run.tcl
@@ -17,6 +17,7 @@ proc main {} {
         "appendonly yes"
         "enable-protected-configs yes"
         "enable-debug-command yes"
+        "loglevel debug"
     }
     run_tests
     cleanup
```

ny0312 commented:
OK. I am able to reproduce the test failure with 1 core, 512MB RAM 64bit FreeBSD.

The failure is Redis getting OOM killed by kernel due to out of swap. In the test, I'm allowing cluster link buffers to grow up to 32MB. That proved to be too much for the FreeBSD test environment used by the daily runs.

I'm currently experimenting with a new cluster link buffer limit in my test. The challenge is that in order to trigger the condition under test, which is a cluster bus link being freed once over the limit, I need to fill up the cluster link buffer. But to fill up the link buffer, I need to first fill up the TCP write buffer in the kernel level, which varies from OS to OS.

I will post a fix PR as soon as I can get a good stable buffer limit.

ny0312 commented:
> also, what about the test failures in the sanitizer runs, like this one:
> https://github.com/redis/redis/runs/4829401183?check_suite_focus=true#step:9:630

This has a different root cause. In the test, I'm assuming as soon as I send a large PUBLISH command to fill up a cluster link, the link will be freed. But in reality the link will only get freed in the next `clusterCron` run whenever that happens. My test is not accounting for this race condition. I will fix it too.


>not sure if this one is related:
> https://github.com/redis/redis/runs/4810931899?check_suite_focus=true#step:9:466

This link has two failures. `test-sanitizer-address (clang) ` and `test-freebsd`

`test-sanitizer-address (clang) `  is not related to my test. `test-freebsd` is related to my test and it should be the same OOM root cause.

oranagra commented:
i didn't look at the test, but i wanna mention that there are other tests in which we struggled to fill output buffers and had to deal with the OS socket buffers swallowing our bytes.

i think these are in:
tests/unit/client-eviction.tcl
tests/unit/maxmemory.tcl
tests/integration/replication-buffer.tcl

IIRC the best approach we had was to pause the destination process, and then gradually fill more and more data until we see it starts piling up, then stop.

ny0312 commented:
@madolson @oranagra @sundb 

PR to fix flaky test "Disconnect link when send buffer limit reached" - https://github.com/redis/redis/pull/10157

Sorry for all the inconvenience. And thanks for the help and advice.

tezc commented:
@ny0312 
Could you please take a look at this as well : https://github.com/redis/redis/runs/4890129923?check_suite_focus=true

```
Testing unit: 24-links.tcl
00:39:15> (init) Restart killed instances: OK
00:39:15> Cluster nodes are reachable: OK
00:39:15> Cluster nodes hard reset: OK
00:39:15> Cluster Join and auto-discovery test: OK
00:39:17> Before slots allocation, all nodes report cluster failure: OK
00:39:17> Create a cluster with two single-node shards: OK
00:39:21> Cluster should start ok: OK
00:39:21> Each node has two links with each peer: FAILED: Expected 19*2 eq 37 (context: type eval line 11 cmd {assert {$num_peers*2 eq $num_links}} proc ::foreach_instance_id)
(Jumping to next unit after error)
```
Don't know about this test but might be related with timing. I increase "cluster-node-timeout" for something else on my branch, I see this test fails here. Just fyi.

ny0312 commented:
@tezc Thanks for reporting the failure. Sorry for the inconvenience. I updated my PR(https://github.com/redis/redis/pull/10157) to fix this test as well.
